Add multithread capabilities to cache warmer

Tickles-BV opened the pull request:
Hi,

This set of changes adds a new option to the performance command called batchsize, this allows you to pass the amount of simultaneous requests we make to servers, greatly speeding up the process of heating up the cache.

We use the RollingCurlX library to accomplish this.

To achieve these changes we have messed around a fair bit with the internals, but as far as I can see everything still functions, though there are probably things I simply didn't test.

Please test and provide feedback so we can improve this functionality.

Tickles-BV committed:
Split up request into batches of right size

Tickles-BV committed:
Hardcode in RollingCurlX and enable multithreading for requests

Tickles-BV committed:
Move RollingCurlX to Hypernode namespace, update table generator to work

Tickles-BV committed:
Fix building up batches when not in multithread mode

Tickles-BV committed:
Replace generateTablesDataForFormat function with new code from generateTablesData

Tickles-BV committed:
Remove unused psr4 autoloader yaml tag

Tickles-BV committed:
Readd require php part to composer.json

vdloo commented on src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php:
```diff
@@ -175,176 +188,256 @@ protected function getEffectiveUrl($url)
+            }
+
+
+            //$requestArray[] = $result['current']['url'];
```
This commented line can be removed right?

Tickles-BV replied:
Yes, I've removed it.

Tickles-BV commented:
A related question, there are two functions to generate the end table for display.
`generateTablesDataForFormat` and `generateTablesData`
The latter being called if the 'format' option has not been set on the commandline.

My question here is that nowhere in the generateTablesDataForFormat function is anything done with the 'format' option. Because I was unsure of what this function did, it is now simply a copy of the generateTablesData. If someone could fill me in on what this function does and why, I would be very appreciative.

Tickles-BV committed:
Remove commented out line

vdloo commented:
```
php /usr/local/bin/magerun2 hypernode:performance --silent --limit 20 --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=http://danieltest2.hypernode.io/
                             
  [Exception]                                                                                                                                                                     
  Notice: Undefined variable: progress in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php on line 397  
```

vdloo commented:
> Because I was unsure of what this function did

it probably does something with the 'output to json' functionality. looks like it was introduced in the commit: https://github.com/Hypernode/hypernode-magerun/commit/b220d3cf7f5eed0e57cdfd828a9a51c97110b2f3 

we do still need to output to json though. If I remember correctly if you pass --silent it will print json

vdloo commented on src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php:
```diff
@@ -175,176 +188,256 @@ protected function getEffectiveUrl($url)
-                }
-                array_push($tableArray['requests'], $requestArray);
+
+            //$requestArray[] = $result['current']['url'];
```
this comment can probably be removed as well

vdloo commented:
```
$ php /usr/local/bin/magerun2 hypernode:performance  --limit 20 --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=http://danieltest2.hypernode.io/
Found 1 batches to process.
 Now executing batch: 1/1
 
 0/20 [>---------------------------]    0% - < 1 sec/< 1 sec 
                                                                                                                                                    
  [Exception]                                                                                                                                       
  Notice: Undefined index: status in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/Performan  
  ceCommand.php on line 456                                           
```

vdloo commented:
```
php /usr/local/bin/magerun2 hypernode:performance --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=http://danieltest2.hypernode.io/

                                                                                                                                                    
  [Exception]                                                                                                                                       
  Warning: explode(): Empty delimiter in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/Perfo  
  rmanceCommand.php on line 677      
```

Tickles-BV committed:
Check if output is silenced before having threads mess with nonexistant progressbar

Tickles-BV committed:
Remove commented out code

Tickles-BV committed:
Return the right response for magerun and magerun2

Tickles-BV committed:
Put $options in the right place so we pass the correct curl options to the processes

Tickles-BV commented:
> $ php /usr/local/bin/magerun2 hypernode:performance  --limit 20 --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=http://danieltest2.hypernode.io/
> Found 1 batches to process.
>  Now executing batch: 1/1
>  
>  0/20 [>---------------------------]    0% - < 1 sec/< 1 sec 
>                                                                                                                                                     
>   [Exception]                                                                                                                                       
>   Notice: Undefined index: status in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/Performan  
>   ceCommand.php on line 456

`magerun2 hypernode:performance --sitemap sitemap.xml --limit 2` Now works fine for me and responds with the correct status

Tickles-BV commented:
> it probably does something with the 'output to json' functionality. looks like it was introduced in the commit: b220d3c
> 
> we do still need to output to json though. If I remember correctly if you pass --silent it will print json

Running `magerun2 hypernode:performance --sitemap sitemap.xml --limit 1 --silent` outputs json now, except the output structure of the json is slightly different (because we changed the internals to support multithreading), the output is now:

`{"\/gear.html":{"current":{"url":"http:\/\/magento2ee.dev\/gear.html","status":200,"ttfb":0.068443}}}`

Where the first key is the url path, the second part is an array that shows if this information is for the current site or the site we're comparing against (second part is empty if --compare-url is not set) 

I'm not sure how critical the json output format is?

Tickles-BV commented:
> php /usr/local/bin/magerun2 hypernode:performance --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=http://danieltest2.hypernode.io/
> 
>                                                                                                                                                     
>   [Exception]                                                                                                                                       
>   Warning: explode(): Empty delimiter in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/Perfo  
>   rmanceCommand.php on line 677

I cannot reproduce this one, could you perhaps share your sitemap.txt?

vdloo commented:
sure
```
$ cat /tmp/hyperformance_sitemap.txt 
http://vdloo.hypernode.io/
http://vdloo.hypernode.io/gear.html
http://vdloo.hypernode.io/gear/bags.html
http://vdloo.hypernode.io/gear/fitness-equipment.html
http://vdloo.hypernode.io/gear/watches.html
http://vdloo.hypernode.io/collections/yoga-new.html
http://vdloo.hypernode.io/training.html
http://vdloo.hypernode.io/training/training-video.html
http://vdloo.hypernode.io/men.html
http://vdloo.hypernode.io/men/tops-men.html
http://vdloo.hypernode.io/men/bottoms-men.html
http://vdloo.hypernode.io/men/tops-men/jackets-men.html
http://vdloo.hypernode.io/men/tops-men/hoodies-and-sweatshirts-men.html
http://vdloo.hypernode.io/men/tops-men/tees-men.html
http://vdloo.hypernode.io/men/tops-men/tanks-men.html
http://vdloo.hypernode.io/men/bottoms-men/pants-men.html
http://vdloo.hypernode.io/men/bottoms-men/shorts-men.html
http://vdloo.hypernode.io/women.html
http://vdloo.hypernode.io/women/tops-women.html
http://vdloo.hypernode.io/women/bottoms-women.html
http://vdloo.hypernode.io/women/tops-women/jackets-women.html
http://vdloo.hypernode.io/women/tops-women/hoodies-and-sweatshirts-women.html
http://vdloo.hypernode.io/women/tops-women/tees-women.html
http://vdloo.hypernode.io/women/tops-women/tanks-women.html
http://vdloo.hypernode.io/women/bottoms-women/pants-women.html
http://vdloo.hypernode.io/women/bottoms-women/shorts-women.html
http://vdloo.hypernode.io/promotions/women-sale.html
http://vdloo.hypernode.io/promotions/men-sale.html
http://vdloo.hypernode.io/promotions/pants-all.html
http://vdloo.hypernode.io/promotions/tees-all.html
http://vdloo.hypernode.io/collections/erin-recommends.html
http://vdloo.hypernode.io/collections/performance-fabrics.html
http://vdloo.hypernode.io/collections/eco-friendly.html
http://vdloo.hypernode.io/sale.html
http://vdloo.hypernode.io/what-is-new.html
http://vdloo.hypernode.io/collections/performance-new.html
http://vdloo.hypernode.io/collections/eco-new.html
http://vdloo.hypernode.io/home
http://vdloo.hypernode.io/enable-cookies
http://vdloo.hypernode.io/privacy-policy-cookie-restriction-mode
http://vdloo.hypernode.io/about-us
http://vdloo.hypernode.io/customer-service
```
note that the sitemap can be either xml or a flat file. this is output from `magerun2 sys:url:list 1 --add-cmspages --add-categories`

vdloo commented:
> I'm not sure how critical the json output format is?

very critical, people have built automation around this so we can't just change it. the format isn't the nicest currently but I'd prefer it if we keep it the same for compatibility

Tickles-BV commented:
> sure
> 
> $ cat /tmp/hyperformance_sitemap.txt 
> http://vdloo.hypernode.io/
> http://vdloo.hypernode.io/gear.html
> http://vdloo.hypernode.io/gear/bags.html
> http://vdloo.hypernode.io/gear/fitness-equipment.html
> http://vdloo.hypernode.io/gear/watches.html
> http://vdloo.hypernode.io/collections/yoga-new.html
> http://vdloo.hypernode.io/training.html
> http://vdloo.hypernode.io/training/training-video.html
> http://vdloo.hypernode.io/men.html
> http://vdloo.hypernode.io/men/tops-men.html
> http://vdloo.hypernode.io/men/bottoms-men.html
> http://vdloo.hypernode.io/men/tops-men/jackets-men.html
> http://vdloo.hypernode.io/men/tops-men/hoodies-and-sweatshirts-men.html
> http://vdloo.hypernode.io/men/tops-men/tees-men.html
> http://vdloo.hypernode.io/men/tops-men/tanks-men.html
> http://vdloo.hypernode.io/men/bottoms-men/pants-men.html
> http://vdloo.hypernode.io/men/bottoms-men/shorts-men.html
> http://vdloo.hypernode.io/women.html
> http://vdloo.hypernode.io/women/tops-women.html
> http://vdloo.hypernode.io/women/bottoms-women.html
> http://vdloo.hypernode.io/women/tops-women/jackets-women.html
> http://vdloo.hypernode.io/women/tops-women/hoodies-and-sweatshirts-women.html
> http://vdloo.hypernode.io/women/tops-women/tees-women.html
> http://vdloo.hypernode.io/women/tops-women/tanks-women.html
> http://vdloo.hypernode.io/women/bottoms-women/pants-women.html
> http://vdloo.hypernode.io/women/bottoms-women/shorts-women.html
> http://vdloo.hypernode.io/promotions/women-sale.html
> http://vdloo.hypernode.io/promotions/men-sale.html
> http://vdloo.hypernode.io/promotions/pants-all.html
> http://vdloo.hypernode.io/promotions/tees-all.html
> http://vdloo.hypernode.io/collections/erin-recommends.html
> http://vdloo.hypernode.io/collections/performance-fabrics.html
> http://vdloo.hypernode.io/collections/eco-friendly.html
> http://vdloo.hypernode.io/sale.html
> http://vdloo.hypernode.io/what-is-new.html
> http://vdloo.hypernode.io/collections/performance-new.html
> http://vdloo.hypernode.io/collections/eco-new.html
> http://vdloo.hypernode.io/home
> http://vdloo.hypernode.io/enable-cookies
> http://vdloo.hypernode.io/privacy-policy-cookie-restriction-mode
> http://vdloo.hypernode.io/about-us
> http://vdloo.hypernode.io/customer-service
> note that the sitemap can be either xml or a flat file. this is output from magerun2 sys:url:list 1 --add-cmspages --add-categories

Could you check if this bug is also present if you use the master branch instead of this one? It is for me but I'd like confirmation.

vdloo commented:
looks like you're right, it's in master as well:
```
app@j6yr83-apenvliegenniet-magweb-cmbl:~/public$ php /usr/local/bin/magerun2 hypernode:performance --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=http://danieltest2.hypernode.io/

                                                                                                                                                    
  [Exception]                                                                                                                                       
  Warning: explode(): Empty delimiter in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/Perfo  
  rmanceCommand.php on line 579     
```
I'll create a separate issue for that

Tickles-BV commented:
> very critical, people have built automation around this so we can't just change it. the format isn't the nicest currently but I'd prefer it if we keep it the same for compatibility

Is there a definition for the format somewhere? I could write a shim that transforms the current internal structure to the old one, but I'd have to know the actual structure.

Basic testing seems to tell me its 3 arrays with the middle one containing all the results in objects, matched by the 'comparison_key' key, but I'm sure there are a bunch of edgecases I'm missing that way.

vdloo commented:
> Is there a definition for the format somewhere?

no definition but I do have an example:
```
[[[{"url":"http:\/\/danieltest2.hypernode.io\/","status":200,"ttfb":0.112313,"comparison_key":"\/"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/","status":200,"ttfb":0.027963,"comparison_key":"\/"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear.html","status":200,"ttfb":0.104849,"comparison_key":"\/gear.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear.html","status":200,"ttfb":0.0283,"comparison_key":"\/gear.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear\/bags.html","status":200,"ttfb":0.111029,"comparison_key":"\/gear\/bags.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear\/bags.html","status":200,"ttfb":0.034397,"comparison_key":"\/gear\/bags.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear\/fitness-equipment.html","status":200,"ttfb":0.103627,"comparison_key":"\/gear\/fitness-equipment.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear\/fitness-equipment.html","status":200,"ttfb":0.033069,"comparison_key":"\/gear\/fitness-equipment.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear\/watches.html","status":200,"ttfb":0.092266,"comparison_key":"\/gear\/watches.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear\/watches.html","status":200,"ttfb":0.033322,"comparison_key":"\/gear\/watches.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/collections\/yoga-new.html","status":200,"ttfb":0.112879,"comparison_key":"\/collections\/yoga-new.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/collections\/yoga-new.html","status":200,"ttfb":0.028833,"comparison_key":"\/collections\/yoga-new.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/training.html","status":200,"ttfb":0.087163,"comparison_key":"\/training.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/training.html","status":200,"ttfb":0.027953,"comparison_key":"\/training.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/training\/training-video.html","status":200,"ttfb":0.106886,"comparison_key":"\/training\/training-video.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/training\/training-video.html","status":200,"ttfb":0.031174,"comparison_key":"\/training\/training-video.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men.html","status":200,"ttfb":0.090725,"comparison_key":"\/men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men.html","status":200,"ttfb":0.038918,"comparison_key":"\/men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men.html","status":200,"ttfb":0.112325,"comparison_key":"\/men\/tops-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men.html","status":200,"ttfb":0.037024,"comparison_key":"\/men\/tops-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/bottoms-men.html","status":200,"ttfb":0.116646,"comparison_key":"\/men\/bottoms-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/bottoms-men.html","status":200,"ttfb":0.034279,"comparison_key":"\/men\/bottoms-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/jackets-men.html","status":200,"ttfb":0.092401,"comparison_key":"\/men\/tops-men\/jackets-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/jackets-men.html","status":200,"ttfb":0.039643,"comparison_key":"\/men\/tops-men\/jackets-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/hoodies-and-sweatshirts-men.html","status":200,"ttfb":0.099496,"comparison_key":"\/men\/tops-men\/hoodies-and-sweatshirts-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/hoodies-and-sweatshirts-men.html","status":200,"ttfb":0.039103,"comparison_key":"\/men\/tops-men\/hoodies-and-sweatshirts-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/tees-men.html","status":200,"ttfb":0.092021,"comparison_key":"\/men\/tops-men\/tees-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/tees-men.html","status":200,"ttfb":0.034356,"comparison_key":"\/men\/tops-men\/tees-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/tanks-men.html","status":200,"ttfb":0.115129,"comparison_key":"\/men\/tops-men\/tanks-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/tanks-men.html","status":200,"ttfb":0.032948,"comparison_key":"\/men\/tops-men\/tanks-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/bottoms-men\/pants-men.html","status":200,"ttfb":0.095455,"comparison_key":"\/men\/bottoms-men\/pants-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/bottoms-men\/pants-men.html","status":200,"ttfb":0.035333,"comparison_key":"\/men\/bottoms-men\/pants-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/bottoms-men\/shorts-men.html","status":200,"ttfb":0.11723,"comparison_key":"\/men\/bottoms-men\/shorts-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/bottoms-men\/shorts-men.html","status":200,"ttfb":0.036524,"comparison_key":"\/men\/bottoms-men\/shorts-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/women.html","status":200,"ttfb":0.10609,"comparison_key":"\/women.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/women.html","status":200,"ttfb":0.037631,"comparison_key":"\/women.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/women\/tops-women.html","status":200,"ttfb":0.105036,"comparison_key":"\/women\/tops-women.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/women\/tops-women.html","status":200,"ttfb":0.035213,"comparison_key":"\/women\/tops-women.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/women\/bottoms-women.html","status":200,"ttfb":0.078326,"comparison_key":"\/women\/bottoms-women.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/women\/bottoms-women.html","status":200,"ttfb":0.038344,"comparison_key":"\/women\/bottoms-women.html"}]]]
```

Tickles-BV committed:
Add shim for json output

Tickles-BV committed:
Code format

Tickles-BV committed:
Remove return types to keep php5 compatibility

Tickles-BV commented:
> Is there a definition for the format somewhere?
> no definition but I do have an example:
> 
> [[[{"url":"http:\/\/danieltest2.hypernode.io\/","status":200,"ttfb":0.112313,"comparison_key":"\/"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/","status":200,"ttfb":0.027963,"comparison_key":"\/"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear.html","status":200,"ttfb":0.104849,"comparison_key":"\/gear.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear.html","status":200,"ttfb":0.0283,"comparison_key":"\/gear.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear\/bags.html","status":200,"ttfb":0.111029,"comparison_key":"\/gear\/bags.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear\/bags.html","status":200,"ttfb":0.034397,"comparison_key":"\/gear\/bags.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear\/fitness-equipment.html","status":200,"ttfb":0.103627,"comparison_key":"\/gear\/fitness-equipment.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear\/fitness-equipment.html","status":200,"ttfb":0.033069,"comparison_key":"\/gear\/fitness-equipment.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/gear\/watches.html","status":200,"ttfb":0.092266,"comparison_key":"\/gear\/watches.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/gear\/watches.html","status":200,"ttfb":0.033322,"comparison_key":"\/gear\/watches.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/collections\/yoga-new.html","status":200,"ttfb":0.112879,"comparison_key":"\/collections\/yoga-new.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/collections\/yoga-new.html","status":200,"ttfb":0.028833,"comparison_key":"\/collections\/yoga-new.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/training.html","status":200,"ttfb":0.087163,"comparison_key":"\/training.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/training.html","status":200,"ttfb":0.027953,"comparison_key":"\/training.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/training\/training-video.html","status":200,"ttfb":0.106886,"comparison_key":"\/training\/training-video.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/training\/training-video.html","status":200,"ttfb":0.031174,"comparison_key":"\/training\/training-video.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men.html","status":200,"ttfb":0.090725,"comparison_key":"\/men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men.html","status":200,"ttfb":0.038918,"comparison_key":"\/men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men.html","status":200,"ttfb":0.112325,"comparison_key":"\/men\/tops-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men.html","status":200,"ttfb":0.037024,"comparison_key":"\/men\/tops-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/bottoms-men.html","status":200,"ttfb":0.116646,"comparison_key":"\/men\/bottoms-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/bottoms-men.html","status":200,"ttfb":0.034279,"comparison_key":"\/men\/bottoms-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/jackets-men.html","status":200,"ttfb":0.092401,"comparison_key":"\/men\/tops-men\/jackets-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/jackets-men.html","status":200,"ttfb":0.039643,"comparison_key":"\/men\/tops-men\/jackets-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/hoodies-and-sweatshirts-men.html","status":200,"ttfb":0.099496,"comparison_key":"\/men\/tops-men\/hoodies-and-sweatshirts-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/hoodies-and-sweatshirts-men.html","status":200,"ttfb":0.039103,"comparison_key":"\/men\/tops-men\/hoodies-and-sweatshirts-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/tees-men.html","status":200,"ttfb":0.092021,"comparison_key":"\/men\/tops-men\/tees-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/tees-men.html","status":200,"ttfb":0.034356,"comparison_key":"\/men\/tops-men\/tees-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/tops-men\/tanks-men.html","status":200,"ttfb":0.115129,"comparison_key":"\/men\/tops-men\/tanks-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/tops-men\/tanks-men.html","status":200,"ttfb":0.032948,"comparison_key":"\/men\/tops-men\/tanks-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/bottoms-men\/pants-men.html","status":200,"ttfb":0.095455,"comparison_key":"\/men\/bottoms-men\/pants-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/bottoms-men\/pants-men.html","status":200,"ttfb":0.035333,"comparison_key":"\/men\/bottoms-men\/pants-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/men\/bottoms-men\/shorts-men.html","status":200,"ttfb":0.11723,"comparison_key":"\/men\/bottoms-men\/shorts-men.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/men\/bottoms-men\/shorts-men.html","status":200,"ttfb":0.036524,"comparison_key":"\/men\/bottoms-men\/shorts-men.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/women.html","status":200,"ttfb":0.10609,"comparison_key":"\/women.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/women.html","status":200,"ttfb":0.037631,"comparison_key":"\/women.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/women\/tops-women.html","status":200,"ttfb":0.105036,"comparison_key":"\/women\/tops-women.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/women\/tops-women.html","status":200,"ttfb":0.035213,"comparison_key":"\/women\/tops-women.html"}],[{"url":"http:\/\/danieltest2.hypernode.io\/women\/bottoms-women.html","status":200,"ttfb":0.078326,"comparison_key":"\/women\/bottoms-women.html"},{"url":"http:\/\/apenvliegenniet.hypernode.io\/women\/bottoms-women.html","status":200,"ttfb":0.038344,"comparison_key":"\/women\/bottoms-women.html"}]]]

I've added a shim called prepareJsonOutput that should transform the array to the old standard.

Sample output:
`[[[{"url":"http:\/\/magento2ee.dev\/producten.html","status":404,"ttfb":0.75266,"comparison_key":"\/producten.html"},{"url":"http:\/\/magento2.dev\/producten.html","status":404,"ttfb":1.091198,"comparison_key":"\/producten.html"}]]]`

vdloo commented:
great! there are still some issues though. the latest revision breaks plaintext sitemap support:
```
php /usr/local/bin/magerun2 hypernode:performance --silent --limit 20 --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=danieltest2.hypernode.io
PHP Warning:  SimpleXMLElement::__construct(): Entity: line 1: parser error : Start tag expected, '<' not found in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php on line 558
PHP Warning:  SimpleXMLElement::__construct(): http://vdloo.hypernode.io/ in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php on line 558
PHP Warning:  SimpleXMLElement::__construct(): ^ in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php on line 558
String could not be parsed as XML /tmp/hyperformance_sitemap.txt
```
it now tries to interpret a list of url as xml:
```
$ cat /tmp/hyperformance_sitemap.txt 
http://vdloo.hypernode.io/
http://vdloo.hypernode.io/gear.html
http://vdloo.hypernode.io/gear/bags.html
http://vdloo.hypernode.io/gear/fitness-equipment.html
http://vdloo.hypernode.io/gear/watches.html
http://vdloo.hypernode.io/collections/yoga-new.html
http://vdloo.hypernode.io/training.html
http://vdloo.hypernode.io/training/training-video.html
http://vdloo.hypernode.io/men.html
```

Tickles-BV commented:
> great! there are still some issues though. the latest revision breaks plaintext sitemap support:
> 
> php /usr/local/bin/magerun2 hypernode:performance --silent --limit 20 --compare-url=vdloo.hypernode.io --sitemap /tmp/hyperformance_sitemap.txt --current-url=danieltest2.hypernode.io
> PHP Warning:  SimpleXMLElement::__construct(): Entity: line 1: parser error : Start tag expected, '<' not found in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php on line 558
> PHP Warning:  SimpleXMLElement::__construct(): http://vdloo.hypernode.io/ in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php on line 558
> PHP Warning:  SimpleXMLElement::__construct(): ^ in /usr/local/share/n98-magerun2/modules/Hypernode/src/Hypernode/Magento/Command/Hypernode/Performance/PerformanceCommand.php on line 558
> String could not be parsed as XML /tmp/hyperformance_sitemap.txt
> it now tries to interpret a list of url as xml:
> 
> $ cat /tmp/hyperformance_sitemap.txt 
> http://vdloo.hypernode.io/
> http://vdloo.hypernode.io/gear.html
> http://vdloo.hypernode.io/gear/bags.html
> http://vdloo.hypernode.io/gear/fitness-equipment.html
> http://vdloo.hypernode.io/gear/watches.html
> http://vdloo.hypernode.io/collections/yoga-new.html
> http://vdloo.hypernode.io/training.html
> http://vdloo.hypernode.io/training/training-video.html
> http://vdloo.hypernode.io/men.html

Hi,

There seem to be various issues with the way sitemaps are generated, though I have not touched this code.

You pass it a url, and it checks if the file exists in the root directory of the magento store, if that is the case it never checks for filetype (or extension) and always tries to load it as an xml file. If it does not exist it tries to load the file by the path you actually passed it, checks if it is a textfile or an xml file and loads it.

This seems wrong, and beyond the scope of my pull request, it is also kind of unrelated because the only way I can replicate this issue is by being outside a magento2 root directory, and not passing it one with --root-dir. Which again seems wrong, but also beyond the scope of my pull request.

vdloo commented:
the structure of the output is still a bit different from what it was before, example:
```
before:
[[[{u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.113904,
    u'url': u'http://danieltest2.hypernode.io/'},
   {u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.030651,
    u'url': u'http://apenvliegenniet.hypernode.io/'}],
```

```
after:
[[[{u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.089545,
    u'url': u'http://danieltest2.hypernode.io/'},
   {u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.054306,
    u'url': u'http://vdloo.hypernode.io/'},
```

vdloo commented:
@Tickles-BV disregard those last comments, I deleted them but github still sent the email probably. you are correct that these are out of the scope of this PR. I'm still reviewing the new changes but overall this looks almost good to go

Tickles-BV commented:
> the structure of the output is still a bit different from what it was before, example:
> 
> before:
> [[[{u'comparison_key': u'/',
>     u'status': 200,
>     u'ttfb': 0.113904,
>     u'url': u'http://danieltest2.hypernode.io/'},
>    {u'comparison_key': u'/',
>     u'status': 200,
>     u'ttfb': 0.030651,
>     u'url': u'http://apenvliegenniet.hypernode.io/'}],
> after:
> [[[{u'comparison_key': u'/',
>     u'status': 200,
>     u'ttfb': 0.089545,
>     u'url': u'http://danieltest2.hypernode.io/'},
>    {u'comparison_key': u'/',
>     u'status': 200,
>     u'ttfb': 0.054306,
>     u'url': u'http://vdloo.hypernode.io/'},

So all the json entities with the same comparison_key should be in the same array? Because that would be an easy fix, but I'm worried I'm still missing something

vdloo commented:
> the structure of the output is still a bit different from what it was before

Looks like I pasted an incomplete sample. The difference is that previously the results for both sites as dicts should be in a list for every url like so:

```
[[[{u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.113904,
    u'url': u'http://danieltest2.hypernode.io/'},
   {u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.030651,
    u'url': u'http://apenvliegenniet.hypernode.io/'}],
  [{u'comparison_key': u'/gear.html',
    u'status': 200,
    u'ttfb': 0.102251,
    u'url': u'http://danieltest2.hypernode.io/gear.html'},
   {u'comparison_key': u'/gear.html',
    u'status': 200,
    u'ttfb': 0.029547,
    u'url': u'http://apenvliegenniet.hypernode.io/gear.html'}],
  [{u'comparison_key': u'/gear/bags.html',
    u'status': 200,
    u'ttfb': 0.116592,
    u'url': u'http://danieltest2.hypernode.io/gear/bags.html'},
   {u'comparison_key': u'/gear/bags.html',
    u'status': 200,
    u'ttfb': 0.03373,
    u'url': u'http://apenvliegenniet.hypernode.io/gear/bags.html'}],
  [{u'comparison_key': u'/gear/fitness-equipment.html',
    u'status': 200,
    u'ttfb': 0.103721,
    u'url': u'http://danieltest2.hypernode.io/gear/fitness-equipment.html'},
   {u'comparison_key': u'/gear/fitness-equipment.html',
    u'status': 200,
    u'ttfb': 0.04059,
    u'url': u'http://apenvliegenniet.hypernode.io/gear/fitness-equipment.html'}],
  [{u'comparison_key': u'/gear/watches.html',
    u'status': 200,
    u'ttfb': 0.104597,
    u'url': u'http://danieltest2.hypernode.io/gear/watches.html'},
   {u'comparison_key': u'/gear/watches.html',
    u'status': 200,
    u'ttfb': 0.034937,
    u'url': u'http://apenvliegenniet.hypernode.io/gear/watches.html'}],
  [{u'comparison_key': u'/collections/yoga-new.html',
    u'status': 200,
    u'ttfb': 0.117983,
    u'url': u'http://danieltest2.hypernode.io/collections/yoga-new.html'},
   {u'comparison_key': u'/collections/yoga-new.html',
    u'status': 200,
    u'ttfb': 0.041197,
    u'url': u'http://apenvliegenniet.hypernode.io/collections/yoga-new.html'}],
  [{u'comparison_key': u'/training.html',
    u'status': 200,
    u'ttfb': 0.12883,
    u'url': u'http://danieltest2.hypernode.io/training.html'},
   {u'comparison_key': u'/training.html',
    u'status': 200,
    u'ttfb': 0.09444,
    u'url': u'http://apenvliegenniet.hypernode.io/training.html'}],
  [{u'comparison_key': u'/training/training-video.html',
    u'status': 200,
    u'ttfb': 0.097972,
    u'url': u'http://danieltest2.hypernode.io/training/training-video.html'},
   {u'comparison_key': u'/training/training-video.html',
    u'status': 200,
    u'ttfb': 0.056902,
    u'url': u'http://apenvliegenniet.hypernode.io/training/training-video.html'}],
  [{u'comparison_key': u'/men.html',
    u'status': 200,
    u'ttfb': 0.117059,
    u'url': u'http://danieltest2.hypernode.io/men.html'},
   {u'comparison_key': u'/men.html',
    u'status': 200,
    u'ttfb': 0.155534,
    u'url': u'http://apenvliegenniet.hypernode.io/men.html'}],
  [{u'comparison_key': u'/men/tops-men.html',
    u'status': 200,
    u'ttfb': 0.11311,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men.html'},
   {u'comparison_key': u'/men/tops-men.html',
    u'status': 200,
    u'ttfb': 0.214571,
    u'url': u'http://apenvliegenniet.hypernode.io/men/tops-men.html'}],
  [{u'comparison_key': u'/men/bottoms-men.html',
    u'status': 200,
    u'ttfb': 0.110959,
    u'url': u'http://danieltest2.hypernode.io/men/bottoms-men.html'},
   {u'comparison_key': u'/men/bottoms-men.html',
    u'status': 200,
    u'ttfb': 0.178092,
    u'url': u'http://apenvliegenniet.hypernode.io/men/bottoms-men.html'}],
  [{u'comparison_key': u'/men/tops-men/jackets-men.html',
    u'status': 200,
    u'ttfb': 0.112548,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/jackets-men.html'},
   {u'comparison_key': u'/men/tops-men/jackets-men.html',
    u'status': 200,
    u'ttfb': 0.030844,
    u'url': u'http://apenvliegenniet.hypernode.io/men/tops-men/jackets-men.html'}],
  [{u'comparison_key': u'/men/tops-men/hoodies-and-sweatshirts-men.html',
    u'status': 200,
    u'ttfb': 0.102828,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/hoodies-and-sweatshirts-men.html'},
   {u'comparison_key': u'/men/tops-men/hoodies-and-sweatshirts-men.html',
    u'status': 200,
    u'ttfb': 0.031203,
    u'url': u'http://apenvliegenniet.hypernode.io/men/tops-men/hoodies-and-sweatshirts-men.html'}],
  [{u'comparison_key': u'/men/tops-men/tees-men.html',
    u'status': 200,
    u'ttfb': 0.125236,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/tees-men.html'},
   {u'comparison_key': u'/men/tops-men/tees-men.html',
    u'status': 200,
    u'ttfb': 0.032071,
    u'url': u'http://apenvliegenniet.hypernode.io/men/tops-men/tees-men.html'}],
  [{u'comparison_key': u'/men/tops-men/tanks-men.html',
    u'status': 200,
    u'ttfb': 0.106262,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/tanks-men.html'},
   {u'comparison_key': u'/men/tops-men/tanks-men.html',
    u'status': 200,
    u'ttfb': 0.062797,
    u'url': u'http://apenvliegenniet.hypernode.io/men/tops-men/tanks-men.html'}],
  [{u'comparison_key': u'/men/bottoms-men/pants-men.html',
    u'status': 200,
    u'ttfb': 0.130403,
    u'url': u'http://danieltest2.hypernode.io/men/bottoms-men/pants-men.html'},
   {u'comparison_key': u'/men/bottoms-men/pants-men.html',
    u'status': 200,
    u'ttfb': 0.103756,
    u'url': u'http://apenvliegenniet.hypernode.io/men/bottoms-men/pants-men.html'}],
  [{u'comparison_key': u'/men/bottoms-men/shorts-men.html',
    u'status': 200,
    u'ttfb': 0.129586,
    u'url': u'http://danieltest2.hypernode.io/men/bottoms-men/shorts-men.html'},
   {u'comparison_key': u'/men/bottoms-men/shorts-men.html',
    u'status': 200,
    u'ttfb': 0.166833,
    u'url': u'http://apenvliegenniet.hypernode.io/men/bottoms-men/shorts-men.html'}],
  [{u'comparison_key': u'/women.html',
    u'status': 200,
    u'ttfb': 0.100346,
    u'url': u'http://danieltest2.hypernode.io/women.html'},
   {u'comparison_key': u'/women.html',
    u'status': 200,
    u'ttfb': 0.17056,
    u'url': u'http://apenvliegenniet.hypernode.io/women.html'}],
  [{u'comparison_key': u'/women/tops-women.html',
    u'status': 200,
    u'ttfb': 0.126391,
    u'url': u'http://danieltest2.hypernode.io/women/tops-women.html'},
   {u'comparison_key': u'/women/tops-women.html',
    u'status': 200,
    u'ttfb': 0.062897,
    u'url': u'http://apenvliegenniet.hypernode.io/women/tops-women.html'}],
  [{u'comparison_key': u'/women/bottoms-women.html',
    u'status': 200,
    u'ttfb': 0.116072,
    u'url': u'http://danieltest2.hypernode.io/women/bottoms-women.html'},
   {u'comparison_key': u'/women/bottoms-women.html',
    u'status': 200,
    u'ttfb': 0.101667,
    u'url': u'http://apenvliegenniet.hypernode.io/women/bottoms-women.html'}]]]
```

now after this PR it is 1 flat list with each result as a dict in that list for all urls:
```
[[[{u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.089545,
    u'url': u'http://danieltest2.hypernode.io/'},
   {u'comparison_key': u'/',
    u'status': 200,
    u'ttfb': 0.054306,
    u'url': u'http://vdloo.hypernode.io/'},
   {u'comparison_key': u'/gear.html',
    u'status': 200,
    u'ttfb': 0.091198,
    u'url': u'http://danieltest2.hypernode.io/gear.html'},
   {u'comparison_key': u'/gear.html',
    u'status': 200,
    u'ttfb': 0.053966,
    u'url': u'http://vdloo.hypernode.io/gear.html'},
   {u'comparison_key': u'/gear/bags.html',
    u'status': 200,
    u'ttfb': 0.100562,
    u'url': u'http://danieltest2.hypernode.io/gear/bags.html'},
   {u'comparison_key': u'/gear/bags.html',
    u'status': 200,
    u'ttfb': 0.054657,
    u'url': u'http://vdloo.hypernode.io/gear/bags.html'},
   {u'comparison_key': u'/gear/fitness-equipment.html',
    u'status': 200,
    u'ttfb': 0.089195,
    u'url': u'http://danieltest2.hypernode.io/gear/fitness-equipment.html'},
   {u'comparison_key': u'/gear/fitness-equipment.html',
    u'status': 200,
    u'ttfb': 0.039154,
    u'url': u'http://vdloo.hypernode.io/gear/fitness-equipment.html'},
   {u'comparison_key': u'/gear/watches.html',
    u'status': 200,
    u'ttfb': 0.07349,
    u'url': u'http://danieltest2.hypernode.io/gear/watches.html'},
   {u'comparison_key': u'/gear/watches.html',
    u'status': 200,
    u'ttfb': 0.065371,
    u'url': u'http://vdloo.hypernode.io/gear/watches.html'},
   {u'comparison_key': u'/collections/yoga-new.html',
    u'status': 200,
    u'ttfb': 0.074022,
    u'url': u'http://danieltest2.hypernode.io/collections/yoga-new.html'},
   {u'comparison_key': u'/collections/yoga-new.html',
    u'status': 200,
    u'ttfb': 0.056787,
    u'url': u'http://vdloo.hypernode.io/collections/yoga-new.html'},
   {u'comparison_key': u'/training.html',
    u'status': 200,
    u'ttfb': 0.071902,
    u'url': u'http://danieltest2.hypernode.io/training.html'},
   {u'comparison_key': u'/training.html',
    u'status': 200,
    u'ttfb': 0.039858,
    u'url': u'http://vdloo.hypernode.io/training.html'},
   {u'comparison_key': u'/training/training-video.html',
    u'status': 200,
    u'ttfb': 0.068789,
    u'url': u'http://danieltest2.hypernode.io/training/training-video.html'},
   {u'comparison_key': u'/training/training-video.html',
    u'status': 200,
    u'ttfb': 0.057829,
    u'url': u'http://vdloo.hypernode.io/training/training-video.html'},
   {u'comparison_key': u'/men.html',
    u'status': 200,
    u'ttfb': 0.070906,
    u'url': u'http://danieltest2.hypernode.io/men.html'},
   {u'comparison_key': u'/men.html',
    u'status': 200,
    u'ttfb': 0.042109,
    u'url': u'http://vdloo.hypernode.io/men.html'},
   {u'comparison_key': u'/men/tops-men.html',
    u'status': 200,
    u'ttfb': 0.0708,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men.html'},
   {u'comparison_key': u'/men/tops-men.html',
    u'status': 200,
    u'ttfb': 0.039658,
    u'url': u'http://vdloo.hypernode.io/men/tops-men.html'},
   {u'comparison_key': u'/men/bottoms-men.html',
    u'status': 200,
    u'ttfb': 0.071871,
    u'url': u'http://danieltest2.hypernode.io/men/bottoms-men.html'},
   {u'comparison_key': u'/men/bottoms-men.html',
    u'status': 200,
    u'ttfb': 0.038867,
    u'url': u'http://vdloo.hypernode.io/men/bottoms-men.html'},
   {u'comparison_key': u'/men/tops-men/jackets-men.html',
    u'status': 200,
    u'ttfb': 0.069779,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/jackets-men.html'},
   {u'comparison_key': u'/men/tops-men/jackets-men.html',
    u'status': 200,
    u'ttfb': 0.042323,
    u'url': u'http://vdloo.hypernode.io/men/tops-men/jackets-men.html'},
   {u'comparison_key': u'/men/tops-men/hoodies-and-sweatshirts-men.html',
    u'status': 200,
    u'ttfb': 0.071236,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/hoodies-and-sweatshirts-men.html'},
   {u'comparison_key': u'/men/tops-men/hoodies-and-sweatshirts-men.html',
    u'status': 200,
    u'ttfb': 0.062632,
    u'url': u'http://vdloo.hypernode.io/men/tops-men/hoodies-and-sweatshirts-men.html'},
   {u'comparison_key': u'/men/tops-men/tees-men.html',
    u'status': 200,
    u'ttfb': 0.073135,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/tees-men.html'},
   {u'comparison_key': u'/men/tops-men/tees-men.html',
    u'status': 200,
    u'ttfb': 0.042097,
    u'url': u'http://vdloo.hypernode.io/men/tops-men/tees-men.html'},
   {u'comparison_key': u'/men/tops-men/tanks-men.html',
    u'status': 200,
    u'ttfb': 0.073056,
    u'url': u'http://danieltest2.hypernode.io/men/tops-men/tanks-men.html'},
   {u'comparison_key': u'/men/tops-men/tanks-men.html',
    u'status': 200,
    u'ttfb': 0.039694,
    u'url': u'http://vdloo.hypernode.io/men/tops-men/tanks-men.html'},
   {u'comparison_key': u'/men/bottoms-men/pants-men.html',
    u'status': 200,
    u'ttfb': 0.070167,
    u'url': u'http://danieltest2.hypernode.io/men/bottoms-men/pants-men.html'},
   {u'comparison_key': u'/men/bottoms-men/pants-men.html',
    u'status': 200,
    u'ttfb': 0.057872,
    u'url': u'http://vdloo.hypernode.io/men/bottoms-men/pants-men.html'},
   {u'comparison_key': u'/men/bottoms-men/shorts-men.html',
    u'status': 200,
    u'ttfb': 0.076479,
    u'url': u'http://danieltest2.hypernode.io/men/bottoms-men/shorts-men.html'},
   {u'comparison_key': u'/men/bottoms-men/shorts-men.html',
    u'status': 200,
    u'ttfb': 0.055572,
    u'url': u'http://vdloo.hypernode.io/men/bottoms-men/shorts-men.html'},
   {u'comparison_key': u'/women.html',
    u'status': 200,
    u'ttfb': 0.067948,
    u'url': u'http://danieltest2.hypernode.io/women.html'},
   {u'comparison_key': u'/women.html',
    u'status': 200,
    u'ttfb': 0.038994,
    u'url': u'http://vdloo.hypernode.io/women.html'},
   {u'comparison_key': u'/women/tops-women.html',
    u'status': 200,
    u'ttfb': 0.099754,
    u'url': u'http://danieltest2.hypernode.io/women/tops-women.html'},
   {u'comparison_key': u'/women/tops-women.html',
    u'status': 200,
    u'ttfb': 0.040624,
    u'url': u'http://vdloo.hypernode.io/women/tops-women.html'},
   {u'comparison_key': u'/women/bottoms-women.html',
    u'status': 200,
    u'ttfb': 0.109944,
    u'url': u'http://danieltest2.hypernode.io/women/bottoms-women.html'},
   {u'comparison_key': u'/women/bottoms-women.html',
    u'status': 200,
    u'ttfb': 0.04028,
    u'url': u'http://vdloo.hypernode.io/women/bottoms-women.html'}]]]
```

Tickles-BV committed:
Change json shim to correct output

Tickles-BV committed:
Add edgecases to json generation

Tickles-BV committed:
Add documentation and use (hopefully) the right comparison key tags in each situation

Tickles-BV commented:
> Looks like I pasted an incomplete sample. The difference is that previously the results for both sites as dicts should be in a list for every url like so:

I've changed the shim so they are dropped in the same array now, I've also added a bunch of logic to replicate the old way of setting the comparison_key, I hope everything works now

vdloo commented:
Very nice, works like a charm :+1: thanks for the contribution

vdloo commented:
I've tagged a release uploaded the new version to the Hypernode repositories:

```
$ dpkg -l | grep magerun-hypernode
ii  magerun-hypernode                     20171011.105637                       all          Magerun modules installed on the Hypernode platform
```

This change will become available on all Hypernodes gradually over the course of this week: https://support.hypernode.com/changelog/release-4475-multi-threaded-magerun-hypernode-cache-warmer-php-update/
